src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Extensions/ContentManagerExtensions.cs

```diff
@@ -0,0 +1,140 @@
+        }
+
+        return results;
+    }
+
+    public static async IAsyncEnumerable<ContentItem> LoadAsync(this IContentManager contentManager, IAsyncEnumerable<ContentItem> contentItems)
+    {
+        await foreach (var contentItem in contentItems)
+        {
+            yield return await contentManager.LoadAsync(contentItem);
+        }
+    }
+
+    public static async Task<ContentValidateResult> UpdateValidateAndCreateAsync(this IContentManager contentManager, ContentItem contentItem, VersionOptions options)
+    {
+        await contentManager.UpdateAsync(contentItem);
+        var result = await contentManager.ValidateAsync(contentItem);
+
+        if (result.Succeeded)
+        {
+            await contentManager.CreateAsync(contentItem, options);
+        }
+
+        return result;
+    }
+
+    /// <summary>
+    /// Gets either the published container content item with the specified id, or if the json path supplied gets the contained content item. 
+    /// </summary>
+    /// <param name="contentManager">The <see cref="IContentManager"/> instance.</param>
+    /// <param name="id">The content item id to load</param>
+    /// <param name="jsonPath">The json path of the contained content item</param>
+    public static Task<ContentItem> GetAsync(this IContentManager contentManager, string id, string jsonPath)
+    {
+        return contentManager.GetAsync(id, jsonPath, VersionOptions.Published);
+    }
+
+    /// <summary>
+    /// Gets either the container content item with the specified id and version, or if the json path supplied gets the contained content item.
+    /// </summary>
+    /// <param name="contentManager">The <see cref="IContentManager"/> instance.</param>
+    /// <param name="id">The id content item id to load</param>
+    /// <param name="options">The version option</param>
+    /// <param name="jsonPath">The json path of the contained content item</param>
+    public static async Task<ContentItem> GetAsync(this IContentManager contentManager, string id, string jsonPath, VersionOptions options)
+    {
+        var contentItem = await contentManager.GetAsync(id, options);
+
+        // It represents a contained content item
+        if (!string.IsNullOrEmpty(jsonPath))
+        {
+            var root = contentItem.Content as JObject;
+            contentItem = root.SelectToken(jsonPath)?.ToObject<ContentItem>();
+
+            return contentItem;
+        }
+
+        return contentItem;
+    }
+
+    /// <summary>
+    /// Gets the content items with the specified ids and version.
+    /// </summary>
+    /// <param name="contentManager">The <see cref="IContentManager"/>.</param>
+    /// <param name="contentItemIds">The content item ids to load.</param>
+    /// <param name="options">The version options.</param>
+    public static async Task<IEnumerable<ContentItem>> GetAsync(this IContentManager contentManager, IEnumerable<string> contentItemIds, VersionOptions options)
```

Comment:
@hishamco This is not a good implementation. If you pass 100 ids, this will query the database 100 times if you use `IsDraftRequired` it potentially will create 100 more queries if the provides ids do not have `latest version` in the database.

The implementation should look more like the following. note I have not tried the code. Also, you need to implement a new `BuildNewVersionsAsync` method similar to the existing `BuildNewVersionAsync()`
```
public async Task<ContentItem> GetAsync(IEnumerable<string> contentItemIds, VersionOptions options)
{
    var contentItems = new List<ContentItem>();
    if (options.IsLatest)
    {
        contentItems = (await _session
            .Query<ContentItem, ContentItemIndex>()
            .Where(x => x.ContentItemId.IsIn(contentItemIds) && x.Latest == true)
            .ListAsync()).ToList();
    }
    else if (options.IsDraft && !options.IsDraftRequired)
    {
        contentItems = (await _session
            .Query<ContentItem, ContentItemIndex>()
            .Where(x =>
                x.ContentItemId.IsIn(contentItemIds) &&
                x.Published == false &&
                x.Latest == true)
            .ListAsync()).ToList();
    }
    else if (options.IsDraft || options.IsDraftRequired)
    {
        // Loaded whatever is the latest as it will be cloned
        contentItems = (await _session
            .Query<ContentItem, ContentItemIndex>()
            .Where(x =>
                x.ContentItemId.IsIn(contentItemIds) &&
                x.Latest == true)
            .ListAsync()).ToList();
    }
    else if (options.IsPublished)
    {
        var missingIds = new List<string>();

        foreach(var contentItemId in contentItemIds)
        {
            if (_contentManagerSession.RecallPublishedItemId(contentItemId, out contentItem))
            {
                contentItems.Add(contentItem);
            }
            else
            {
                missingIds.Add(contentItemId);
            }
        }

        if(missingIds.Count > 0)
        {
            var missingItems = await _session.ExecuteQuery(new PublishedContentItemById(missingIds)).ListAsync();

            contentItems.AddRange(missingItems);
        }
    }

    var needVersions = new List<ContentItem>();

    foreach(var contentItem in contentItems)
    {
        contentItem = await LoadAsync(contentItem);

        if (options.IsDraftRequired)
        {
            // When draft is required and latest is published a new version is added
            if (contentItem.Published)
            {
                // We save the previous version further because this call might do a session query.
                var contentTypeDefinition = _contentDefinitionManager.GetTypeDefinition(contentItem.ContentType);

                // Check if not versionable, meaning we use only one version
                if (contentTypeDefinition != null && !contentTypeDefinition.IsVersionable())
                {
                    contentItem.Published = false;
                }
                else
                {
                    needVersions.Add(contentItem);
                    continue;
                }
            }

            // Save the new version
            _session.Save(contentItem, checkConcurrency: true);
        }
    }
    if(needVersions.Count > 0)
    {
        var items = await BuildNewVersionsAsync(needVersions);

        foreach(var item in items)
        {
            _session.Save(item, checkConcurrency: true);
        }
    }

    return contentItems;
}
```

Reply:
@MikeAlhayek please hold on, I did some changes locally (not published yet), hope to update the PR